### ISSUE TITLE
Excessive CPU memory need for fine-tuning with long context

### ISSUE THREAD

**Issue by leoleoasd** [NVIDIA/NeMo#14526]
**Describe the bug**

Nemo will try to load the entire dataset in every worker. I created a dummy dataset with about 64k tokens and with num_workers=8 it takes more than 2TBs of memory to load them.

**Steps/Code to reproduce bug**

Load a long dataset with SFT dataset.

**Expected behavior**

Maybe only preprocess the dataset on local_rank=0 then broadcast it?

**Comment by leoleoasd** [comment]
Each worker takes about 40GB of memory to load a `1.2M` dummy data. And it takes ~20 seconds to tokenize it, and due to CPU memory constraints I cannot use more workers.

**Comment by leoleoasd** [comment]
More analysis: The bottleneck is not tokenization, it's attention mask calculation. Enabling `get_attention_mask_from_fusion` resolves this.
